periflow/client/base.py

```diff
@@ -111,7 +96,53 @@ def copy(self) -> "URLTemplate":
         return URLTemplate(pattern=Template(self.pattern.template))
 
 
-class Client(ABC, Generic[T]):
+class RequestInterface:
+    """Request API mixin."""
+
+    def check_request(self, func: Callable[..., Response]) -> Callable[..., Response]:
+        """Wrapper function to send requests with the retry and error handling."""
+
+        @wraps(func)
+        def wrapper(*args, **kwargs) -> Response:  # type: ignore
+            final_exc = None
+            for i in range(API_REQUEST_MAX_RETRIES):
+                try:
+                    return func(*args, **kwargs)
+                except RETRYABLE_REQUEST_ERRORS:
+                    logger.info(
+                        ("Retry the failed request (attempt %s / %s)."),
+                        i + 1,
+                        API_REQUEST_MAX_RETRIES,
+                    )
+            raise MaxRetriesExceededError(final_exc)
+
+        return wrapper
+
+    def paginated_get(
+        self,
+        callback: Callable[..., Response],
+        path: Optional[str] = None,
+        limit: int = DEFAULT_PAGINATION_SIZE,
+        params: Optional[dict[str, Any]] = None,
+    ) -> List[Dict[str, Any]]:
+        """List objects with pagination."""
+        page_size = min(DEFAULT_PAGINATION_SIZE, limit)
+        params = {"limit": page_size, **(params or {})}
+        response_dict = callback(path=path, params=params).json()
+        items = response_dict["results"]
+        next_cursor = response_dict["next_cursor"]
+
+        while next_cursor is not None and len(items) < limit:
+            response_dict = callback(
+                path=path, params={**params, "cursor": next_cursor}
+            ).json()
+            items.extend(response_dict["results"])
+            next_cursor = response_dict["next_cursor"]
+
+        return items
```

Comment:
I am sorry but I cannot understand what you pointed out.
When the `limit` argument is `30`, `page_size` will become `20` and the `callback` will be called twice inside the `while` loop. At the first call, `len(item)` will become `20`, and it will become `30` at the second call.

BTW, this method will be deleted (I forgot to remove this.) and the logic in here has already moved into `Client > list` method.

Reply:
Since this method will be deleted and the logic here has already moved into `Client > list` method, I will make the change at there.